src/services/blocks/trace/Trace.ts

```diff
@@ -0,0 +1,690 @@
+	PActionEvent,
+	PalletKeyInfo,
+	PEvent,
+	Phase,
+	PhaseOther,
+	SpanWithChildren,
+	SpecialKeyInfo,
+	TraceSpan,
+} from './types';
+
+const EMPTY_KEY_INFO = {
+	error: 'key not found',
+};
+
+/**
+ * Known spans
+ */
+const SPANS = {
+	executeBlock: {
+		name: 'execute_block',
+		target: 'frame_executive',
+	},
+	applyExtrinsic: {
+		name: 'apply_extrinsic',
+		target: 'frame_executive',
+	},
+	onInitialize: {
+		name: 'on_initialize',
+		// Targets vary by pallet
+	},
+	onFinalize: {
+		name: 'on_finalize',
+		// Targets vary by pallet
+	},
+};
+
+/**
+ * Mapping of span id => span with children ids
+ */
+type SpansById = Map<number, SpanWithChildren>;
+
+/**
+ * Extrinsic indexs decoded from storage read events, keyed by the span Id to which
+ * the original storage event belonged to.
+ */
+type ExtrinsicIndexBySpanId = Map<number, BN>;
+
+/**
+ * Mapping of span id => array of the events belongin to the span.
+ */
+type EventsByParentId = Map<number, PEvent[]>;
+
+type AccountEventsByAddress = Map<string, PAccountEvent[]>;
+
+// In the future this will need a more chain specific solution that creates
+// keys directly from metadata. https://github.com/paritytech/substrate-api-sidecar/issues/528
+/**
+ * Get all the storage key names based on the ones built into `api`.
+ *
+ * @param api ApiPromise
+ */
+function getKeyNames(api: ApiPromise): Record<string, KeyInfo> {
+	const moduleKeys = Object.keys(api.query).reduce((acc, mod) => {
+		Object.keys(api.query[mod]).forEach((item) => {
+			const queryObj = api.query[mod][item];
+			let key;
+			try {
+				// `slice(2)` is to slice off '0x' prefix
+				key = queryObj.key().slice(2);
+			} catch {
+				key = queryObj.keyPrefix().slice(2);
+			}
+
+			acc[key] = {
+				module: mod,
+				item,
+			};
+		});
+
+		return acc;
+	}, {} as Record<string, KeyInfo>);
+
+	const wellKnownKeys = {
+		'3a636f6465': {
+			special: ':code',
+		},
+		'3a686561707061676573': {
+			special: ':heappages',
+		},
+		'3a65787472696e7369635f696e646578': {
+			special: ':extrinsic_index',
+		},
+		'3a6368616e6765735f74726965': {
+			special: ':changes_trie',
+		},
+		'3a6368696c645f73746f726167653a': {
+			special: ':child_storage:',
+		},
+	};
+
+	return { ...moduleKeys, ...wellKnownKeys };
+}
+
+/**
+ * Parses spans by
+ * 1) creating a Map of span's `id` to the the span with its children
+ * 2) creating an array with the same spans
+ *
+ * @param spans spans to parse
+ */
+function parseSpans(
+	spans: TraceSpan[]
+): {
+	spansById: SpansById;
+	parsedSpans: SpanWithChildren[];
+	executeBlockSpanId: number;
+} {
+	// IMPORTANT for downstream logic: Make sure spans are sorted.
+	// If this gets taken out, then **all** `Operation`s will need to be sorted on
+	// `eventIndex` to ensure integrity in there ordering. (Ordering integrity
+	// is not strictly neccesary for MVP balance reconciliation, but operating
+	// under the assumption it may be neccesary for closer audits.)
+	spans.sort((a, b) => a.id - b.id);
```

Comment:
Realizing we probably don't need to do this. Instead we can avoid sorting here and sorting the events of an action group if we just sort the operations by eventIndex before returning.

The current strategy is sort spans, so actionGroups are in the order of there primary span. And sort events within an actionGroup so Operations within an actionGroup are in order. We merge together operations in the order of action groups so it ends up all sorted.